### PR TITLE
Move `updateInputStream` to `ITransformingStep`

### LINKED ISSUE #38116: Filter push down should remove filter columns accurately
**Describe what's wrong**

see subj

**Does it reproduce on recent release?**

yeap

**How to reproduce**

``` SQL
create table t(a UInt8) engine=MergeTree order by a;
insert into t select * from numbers(10);

9d9dd0b8c094 :) select a from t t1 join t t2 on t1.a=t2.a where t1.a                                                                                                                                                                                                             
Code: 10. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Not found column a in block. There are only columns: a. (NOT_FOUND_COLUMN_IN_BLOCK)

9d9dd0b8c094 :) select a from t t1 join t t2 on t1.a=t2.a where t1.a settings query_plan_enable_optimizations=0

┌─a─┐
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
└───┘
```

**Additional context**

Column `t1.a` removed in filter.

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

These changes are needed for query plan optimizations evolving non-adjacent query plan nodes. In this case, we need to call `updateInputStream` on all nodes between them and it is much simpler to do when you don't need to know the actual type of a node. It seems enough to provide this method only for `ITransformingStep` descendants, not for all existing `IQueryPlanStep`-s.

### CODE REVIEW

**Comment by nickitat** [comment]
@Mergifyio update

File: src/Processors/QueryPlan/MergingAggregatedStep.h [hunk 1]
```diff
@@ -27,6 +27,8 @@ class MergingAggregatedStep : public ITransformingStep
     void describeActions(FormatSettings & settings) const override;
 
 private:
+    void updateOutputStream() override;
+
     AggregatingTransformParamsPtr params;
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/MergingAggregatedStep.h:32]: Note that aggregation keys (and aggregates) from params can store *positions*.
If we implement such a method, we must check specified input stream was valid.
I think for some cases we can leave the method "Not implemented"

**Review by KochetovNicolai — APPROVED** [review]
Generally ok

**Comment by nickitat** [comment]
@Megifyio update

**Comment by nickitat** [comment]
@Mergifyio update

**Comment by nickitat** [comment]
@Mergifyio update

File: src/Processors/QueryPlan/Optimizations/liftUpArrayJoin.cpp [hunk 2]
```diff
@@ -36,30 +36,6 @@ size_t tryLiftUpArrayJoin(QueryPlan::Node * parent_node, QueryPlan::Nodes & node
 
     auto description = parent->getStepDescription();
 
-    /// All actions was moved before ARRAY JOIN. Swap Expression and ArrayJoin.
-    if (split_actions.second->trivial())
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/Optimizations/liftUpArrayJoin.cpp:40]: Why this is not needed anymore?

**Review by KochetovNicolai — APPROVED** [review]

**Comment by nickitat** [comment]
@Mergifyio update

**Comment by nickitat** [comment]
@Mergifyio update

**Pull request merged**